### PR TITLE
Unsound call of `set_var`

### PR DESCRIPTION
**Pull request by sftse**
There have been multiple attempts to communicate that the pattern `&mut *(&T as *const T as *mut T)` is not sound, as in #1485 .

~~This is the most straightforward modification of the code that passes `cargo miri`.~~ See #1651 for the fix.
To detect the unsoundness, the parallelism must be turned off under miri, otherwise it falls over because of some unrelated issue in `crossbeam-epoch`:
```
MIRIFLAGS='-Zmiri-env-set=TOKENIZERS_PARALLELISM=false' cargo +nightly miri test bpe::trainer::tests::test_train
```

There ~~are two soundness issues~~is one soundness issue fix and one code cleanup in this PR.

The ~~first~~ soundness issue ~~is transforming & to &mut, the second~~ is hidden by the fact that std until recently had a safe `env::set_var` function. This was a breaking change and is as of now an unsafe function to call.
Lastly, some elided lifetimes were making it quite hard to see whether the use of `RefMutContainer` as mentioned in #1612 was sound. It does not seem to be unsound.

### CODE REVIEW

File: tokenizers/src/models/bpe/trainer.rs [hunk 1]
```diff
@@ -8,6 +8,63 @@ use serde::{Deserialize, Serialize};
...
 use std::collections::{BinaryHeap, HashMap, HashSet};
 
+mod unsafevec {
+    use std::cell::UnsafeCell;
+    use std::marker::PhantomData;
+
+    use super::Word;
+
+    pub(super) struct UnsafeVec<'v> {
+        value: UnsafeCell<*mut Word>,
```
Comment (Manishearth) [inline tokenizers/src/models/bpe/trainer.rs:18]: I don't think this UnsafeCell is doing anything here?

**Review by Manishearth — COMMENTED** [review]
not really convinced this abstraction pulls its weight compared to the approach in #1651 , it doesn't  really internalize any safety invariants

File: tokenizers/src/models/bpe/trainer.rs (hunk 1 above)
Comment (sftse) [reply tokenizers/src/models/bpe/trainer.rs:18]: I wasn't certain when I wrote this, does the borrow on self of `fn get_unchecked_mut` not affect anything?
One is not normally permitted to mutate past a &, you can't use a `&&mut T` to mutate `T`, so I suspected one cannot easily mutate `T` behind `&*mut T`. Miri seems fine with it, so I guess this UnsafeCell is not needed and you're correct.

**Comment by ArthurZucker** [comment]
Thanks a lot for opening the PR and contributing either ways! 🤗

File: bindings/python/src/utils/pretokenization.rs [hunk 2]
```diff
@@ -272,10 +270,10 @@ impl DestroyPtr for PyPreTokenizedStringRefMut {
 }
 
 impl PyPreTokenizedStringRefMut {
-    pub fn new(pretok: &mut tk::PreTokenizedString) -> RefMutGuard<Self> {
+    pub fn new(pretok: &mut tk::PreTokenizedString) -> super::RefMutGuard<'_, Self> {
         // SAFETY: This is safe because we return a RefMutGuard here.
         // The compiler will make sure the &mut stays valid as necessary.
-        RefMutGuard::new(Self {
+        super::RefMutGuard::new(Self {
```
Comment (ArthurZucker) [inline bindings/python/src/utils/pretokenization.rs:276]: any reason not to import it from super directly?

**Review by ArthurZucker — APPROVED** [review]
LGTM had no idea set var was unsafe, we might add this mirin check to our CI!

File: bindings/python/src/utils/pretokenization.rs (hunk 2 above)
Comment (sftse) [reply bindings/python/src/utils/pretokenization.rs]: Commit slipped in from a different PR I was working on, it's an immaterial change, undone.

Comment (sftse) [reply bindings/python/src/utils/pretokenization.rs]: When making refactors, I find it helps the readability of changes to have diffs that are as localized as possible. If subsequent commits touch these lines, I might move the imports to be more local.

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗

**Pull request merged**